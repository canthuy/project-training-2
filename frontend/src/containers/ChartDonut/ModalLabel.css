.modal-custom {
  width: 30% !important;
  position: absolute !important;
  top: 5rem;
  left: 1rem;
}
.input-filter {
  width: 100%;
  border: none;
  outline: none;
  border-bottom: 3px solid #3cb4ec;
}

.btn-labels:active, .btn-labels:focus {
    box-shadow: none !important;
}
.label-os {
    padding: 10px 10px 0;
    display: flex;
    justify-content: space-between;
}
.label-os span {
    font-size: 16px
}
.label-os:hover{
    cursor: pointer;
    color: #3cb4ec
}