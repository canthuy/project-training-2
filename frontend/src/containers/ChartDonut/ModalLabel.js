import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getData } from '../../redux/doughnut/actions';

import { Modal, Button } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheck } from '@fortawesome/free-solid-svg-icons';

import './ModalLabel.css';

const ModalLabel = () => {
  const listDevices = ['Android', 'iOS', 'Windows', 'Os X', 'Unknown', 'Linux'];
  const [listLabel, setListLabel] = useState(listDevices);
  const [show, setShow] = useState(false);
  const deviceData = useSelector((state) => state.device.deviceData);
  const { startDate, endDate } = useSelector((state) => state.datepicker);
  const [listChecked, setListChecked] = useState(
    deviceData.map((val) => val.x)
  );
  const dispatch = useDispatch();

  const onClickLabel = (val) => {
    setListChecked((prevList) => {
      const isChecked = prevList.includes(val);
      if (isChecked) {
        return prevList.filter((check) => check !== val);
      }
      return [...prevList, val];
    });
  };

  const onHideModal = () => {
    const prevListChecked = deviceData.map((val) => val.x);
    const isTrue =
      listChecked.length === prevListChecked.length &&
      prevListChecked.every((v) => listChecked.includes(v));

    if (listChecked.length > 0) {
      if (!isTrue) {
        dispatch(getData([startDate, endDate, listChecked]));
      } else {
        setShow(false);
      }
    } else {
      alert('Please select device type');
    }
  };

  const handleFilter = (e) => {
    e.preventDefault();
    const keyFilter = e.target.value.trim();
    if (keyFilter.length > 0) {
      setListLabel((prev) => {
        return prev.filter((val) => val.toLowerCase().includes(keyFilter));
      });
    } else {
      setListLabel(listDevices);
    }
  };

  return (
    <>
      <Button className="btn-labels" size="sm" onClick={() => setShow(true)}>
        Labels
      </Button>
      <Modal
        show={show}
        onHide={onHideModal}
        dialogClassName="modal-custom"
        aria-labelledby="example-custom-modal-styling-title"
      >
        <Modal.Body>
          <input className="input-filter" onChange={handleFilter} />
          {listLabel.map((val, index) => {
            return (
              <div
                key={index}
                className="label-os"
                onClick={() => onClickLabel(val)}
              >
                <span>{val}</span>
                {listChecked.includes(val) && (
                  <FontAwesomeIcon icon={faCheck} data-testid="test-svg" />
                )}
              </div>
            );
          })}
        </Modal.Body>
      </Modal>
    </>
  );
};
export default ModalLabel;
