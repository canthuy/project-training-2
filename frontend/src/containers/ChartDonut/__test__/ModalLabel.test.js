import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import ModalLabel from '../ModalLabel';
window.React = React;

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('ModalLabel component', () => {
  const mockState = {
    device: {
      deviceData: [
        { x: 'Android', y: 20 },
        { x: 'Windows', y: 30 },
        { x: 'iOS', y: 60 },
        { x: 'Os X', y: 40 },
        { x: 'Unknown', y: 10 },
        { x: 'Linux', y: 50 },
      ],
      loading: false,
      isError: false,
    },
    datepicker: {
      startDate: '',
      endDate: '',
    },
  };
  const mockDispatch = jest.fn();
  beforeEach(() => {
    useDispatch.mockImplementation(() => mockDispatch);
    useSelector.mockImplementation((callback) => {
      return callback(mockState);
    });
  });
  afterEach(() => {
    useDispatch.mockClear();
    useSelector.mockClear();
  });

  it('Show modal label', () => {
    const { container, getByText, getByRole } = render(<ModalLabel />);
    expect(container).toMatchSnapshot();
    const buttonLabel = getByRole('button', { name: /Labels/i });
    fireEvent.click(buttonLabel);
    expect(getByText('Android')).toBeTruthy();
    expect(getByText('Labels')).toBeTruthy();
  });
  it('Click label', () => {
    const { getByText, getByRole, queryAllByTestId } = render(<ModalLabel />);
    const buttonLabel = getByRole('button', { name: /Labels/i });
    fireEvent.click(buttonLabel);
    const label = getByText('Android');
    fireEvent.click(label);
    expect(queryAllByTestId('test-svg')).toHaveLength(5);
    fireEvent.click(label);
    expect(queryAllByTestId('test-svg')).toHaveLength(6);
  });
  it('Hide modal label and do not change labels', () => {
    const { getByRole } = render(<ModalLabel />);
    const buttonLabel = getByRole('button', { name: /Labels/i });
    fireEvent.click(buttonLabel);
    fireEvent.click(document.querySelector('.modal-backdrop'));
    expect(mockDispatch).not.toHaveBeenCalled();
  });
  it('Hide modal label and dispatch label', () => {
    const { getByText, getByRole } = render(<ModalLabel />);
    const buttonLabel = getByRole('button', { name: /Labels/i });
    fireEvent.click(buttonLabel);
    fireEvent.click(getByText('Android'));
    fireEvent.click(getByText('iOS'));
    fireEvent.click(document.querySelector('.modal-backdrop'));
    expect(mockDispatch).toHaveBeenCalled();
  });
  it('hide modal and no device type selected', () => {
    global.alert = jest.fn();
    const { getByText, getByRole } = render(<ModalLabel />);
    const buttonLabel = getByRole('button', { name: /Labels/i });
    fireEvent.click(buttonLabel);
    fireEvent.click(getByText('Android'));
    fireEvent.click(getByText('iOS'));
    fireEvent.click(getByText('Linux'));
    fireEvent.click(getByText('Windows'));
    fireEvent.click(getByText('Os X'));
    fireEvent.click(getByText('Unknown'));
    fireEvent.click(document.querySelector('.modal-backdrop'));
    expect(global.alert).toHaveBeenCalled();
  });
  it('Filter input', () => {
    const { queryByText, getByRole } = render(<ModalLabel />);
    const buttonLabel = getByRole('button', { name: /Labels/i });
    fireEvent.click(buttonLabel);
    const inputFilter = getByRole('textbox');
    fireEvent.change(inputFilter, { target: { value: 'android' } });
    expect(inputFilter.value).toEqual('android');
    expect(queryByText('iOS')).not.toBeTruthy();

    fireEvent.change(inputFilter, { target: { value: '' } });
    expect(queryByText('iOS')).toBeTruthy();
  });
});
