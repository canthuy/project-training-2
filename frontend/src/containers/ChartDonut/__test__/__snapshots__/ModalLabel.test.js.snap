// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalLabel component Show modal label 1`] = `
<div>
  <button
    class="btn-labels btn btn-primary btn-sm"
    type="button"
  >
    Labels
  </button>
</div>
`;
